### PR TITLE
Add custom head_dim support to Llama

### PR DESCRIPTION
**Pull request by suhara**
# What does this PR do?

Llama assumes that `head_dim * num_heads == hidden_size` and does not accommodate any models with custom head_dim size. This PR relaxes the assumption and makes Llama use custom head_dim sizes. 

This PR has a dependency on the following PR:

- #32495 
    - `src/transformers/modeling_rope_utils.py` to use `config.head_dim` for RoPE.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -340,23 +340,27 @@ def __init__(self, config: LlamaConfig, layer_idx: Optional[int] = None):
...
-            raise ValueError(
-                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
-                f" and `num_heads`: {self.num_heads})."
-            )
+        if config.head_dim is None:
+            if (self.head_dim * self.num_heads) != self.hidden_size:
+                raise ValueError(
+                    f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
+                    f" and `num_heads`: {self.num_heads})."
+                )
```
Comment (amyeroberts) [inline src/transformers/models/llama/modeling_llama.py:358]:
```suggestion
        if config.head_dim is None and (self.head_dim * self.num_heads) != self.hidden_size:
            raise ValueError(
                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
                f" and `num_heads`: {self.num_heads})."
            )
```

**Comment by amyeroberts** [comment]
cc @ArthurZucker

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (suhara) [reply src/transformers/models/llama/modeling_llama.py]: @amyeroberts Thanks for the suggestion! Updated the if block accordingly.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -340,14 +340,17 @@ def __init__(self, config: LlamaConfig, layer_idx: Optional[int] = None):
         self.attention_dropout = config.attention_dropout
         self.hidden_size = config.hidden_size
         self.num_heads = config.num_attention_heads
-        self.head_dim = self.hidden_size // self.num_heads
+        if config.head_dim is None:
+            self.head_dim = self.hidden_size // self.num_heads
+        else:
+            self.head_dim = config.head_dim
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:346]:
```suggestion
        self.head_dim = config.get("head_dim",  self.hidden_size // self.num_heads)
```

**Review by ArthurZucker — COMMENTED** [review]
Hey! not sure we needs this (what I meant by a regression is that I thought we did allow head dim for llama) other models had this constraint lifted liike gemma I think

**Comment by suhara** [comment]
Hi @ArthurZucker

The motivation is that some custom Llama-architecture based models with custom head_dim sizes cannot be loaded by `LlamaModel` due to this constraint. (Some context: https://github.com/NVIDIA/NeMo/pull/10078)

That said, I understand your concern on the regression issue. What would be your suggestion? If the existing Llama class is supposed to support official Llama models, creating a new class to cover custom Llama-based variant models would be an option?

File: src/transformers/models/llama/configuration_llama.py [hunk 3]
```diff
@@ -187,6 +191,7 @@ def __init__(
         self.attention_bias = attention_bias
         self.attention_dropout = attention_dropout
         self.mlp_bias = mlp_bias
+        self.head_dim = head_dim
```
Comment (ArthurZucker) [inline src/transformers/models/llama/configuration_llama.py:194]: what you can do here is if head_dim is None: `self.head_dim = self.hidden_size // self.num_heads`

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -340,14 +340,14 @@ def __init__(self, config: LlamaConfig, layer_idx: Optional[int] = None):
...
-        self.head_dim = self.hidden_size // self.num_heads
+        self.head_dim = getattr(config, "head_dim",  self.hidden_size // self.num_heads)
         self.num_key_value_heads = config.num_key_value_heads
         self.num_key_value_groups = self.num_heads // self.num_key_value_heads
         self.max_position_embeddings = config.max_position_embeddings
         self.rope_theta = config.rope_theta
         self.is_causal = True
 
-        if (self.head_dim * self.num_heads) != self.hidden_size:
+        if config.head_dim is None and (self.head_dim * self.num_heads) != self.hidden_size:
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:350]:
```suggestion
        if self.head_dim is None and (self.head_dim * self.num_heads) != self.hidden_size:
```

**Review by ArthurZucker — COMMENTED** [review]
Suggestions should fix the CI let's go with this. 
Sorry for the delayed reviewed I was OOO for a bit

File: src/transformers/models/llama/configuration_llama.py (hunk 3 above)
Comment (suhara) [reply src/transformers/models/llama/configuration_llama.py]: Thanks for the suggestion! Added.

**Comment by suhara** [comment]
@ArthurZucker It seems that you created another PR and fix the remaining issues for this. Thank you!

**Comment by suhara** [comment]
Hi @ArthurZucker 

I saw @Qubitium 's message in #32857. The newly created PR is missing the fix for `o_proj`.
I fixed the CI issues nd this PR should be ready to merge. Can you check?

Thanks!

**Comment by suhara** [comment]
Hi @ArthurZucker 

The CI passed. Can you merge this PR (or #32857 after fixing the issue)? Thanks!

**Comment by suhara** [comment]
#32857 has been merged. Close this PR.

**Pull request closed without merging**